File: predict-the-winner/Solution486.java
// https://leetcode.com/problems/predict-the-winner/

import java.util.HashMap;
import java.util.Map;

class Solution486 {

    // T: (O(n^2)), S: (O(n^2))
    public boolean predictTheWinner(int[] nums) {
        var memo = new HashMap<String, Integer>();
        return turn(nums, 0, nums.length - 1, true, memo) >= 0;
    }

    private int turn(int[] nums, int l, int r, boolean player1, Map<String, Integer> memo) {
        String key = l + "," + r + "," + player1;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        if (l > r) {
            return 0;
        }
        int result = 0;
        if (player1) {
            result = Math.max(
                    nums[l] + turn(nums, l + 1, r, false, memo),
                    nums[r] + turn(nums, l, r - 1, false, memo));
        } else {
            result = Math.min(
                    -nums[l] + turn(nums, l + 1, r, true, memo),
                    -nums[r] + turn(nums, l, r - 1, true, memo));
        }
        memo.put(key, result);
        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution486();
        var nums1 = new int[] { 1, 5, 2 };
        var nums2 = new int[] { 1, 5, 233, 7 };
        System.out.println(sol.predictTheWinner(nums1)); // false
        System.out.println(sol.predictTheWinner(nums2)); // true
    }

}

File: predict-the-winner/solution486.py
# https://leetcode.com/problems/predict-the-winner/

from typing import List


class Solution486:

    # T: (O(n^2)), S: (O(n^2))
    def predictTheWinner(self, nums: List[int]) -> bool:
        memo = {}

        def turn(nums, l, r, player1, memo):
            key = (l, r, player1)
            if key in memo:
                return memo[key]
            if l > r:
                return 0

            if player1:
                result = max(
                    nums[l] + turn(nums, l + 1, r, False, memo),
                    nums[r] + turn(nums, l, r - 1, False, memo)
                )
            else:
                result = min(
                    -nums[l] + turn(nums, l + 1, r, True, memo),
                    -nums[r] + turn(nums, l, r - 1, True, memo)
                )

            memo[key] = result
            return result

        return turn(nums, 0, len(nums) - 1, True, memo) >= 0


def main() -> None:
    sol = Solution486()
    nums1 = [1, 5, 2]
    nums2 = [1, 5, 233, 7]
    print(sol.predictTheWinner(nums1))  # False
    print(sol.predictTheWinner(nums2))  # True


if __name__ == "__main__":
    main()
